pkg/cmd/render/bootstrap_ip.go

```diff
@@ -0,0 +1,199 @@
+	ips, err := ipAddrs()
+	if err != nil {
+		return nil, err
+	}
+	addrMap, err := getAddrMap()
+	if err != nil {
+		return nil, err
+	}
+	routeMap, err := getRouteMap()
+	if err != nil {
+		return nil, err
+	}
+
+	addressFilter := AddressFilters(
+		NonDeprecatedAddress,
+		ContainedByCIDR(machineCIDR),
+		AddressNotIn(excludedIPs...),
+	)
+	routeFilter := NonDefaultRoute
+	addresses, err := routableAddresses(addrMap, routeMap, ips, addressFilter, routeFilter)
+	if err != nil {
+		return nil, err
+	}
+	if len(addresses) == 0 {
+		return nil, fmt.Errorf("couldn't any bootstrap node IPs")
+	}
+	if len(addresses) > 1 {
+		klog.Warningf("found multiple candidate bootstrap IPs; using only the first one: %+v", addresses)
+	}
+
+	var bootstrapIP net.IP
+	for _, ip := range addresses {
+		// IPv6
+		if ipv6 && ip.To4() == nil {
+			bootstrapIP = ip
+			break
+		}
+		// IPv4
+		if !ipv6 && ip.To4() != nil {
+			bootstrapIP = ip
+			break
+		}
+	}
+	if bootstrapIP == nil {
+		return nil, fmt.Errorf("couldn't find a suitable bootstrap node IP from candidates: %+v", addresses)
+	}
+
+	return bootstrapIP, nil
+}
+
+// AddressFilter is a function type to filter addresses
+type AddressFilter func(netlink.Addr) bool
+
+// RouteFilter is a function type to filter routes
+type RouteFilter func(netlink.Route) bool
+
+// NonDeprecatedAddress returns true if the address is IPv6 and has a preferred lifetime of 0
+func NonDeprecatedAddress(addr netlink.Addr) bool {
+	return !(net.IPv6len == len(addr.IP) && addr.PreferedLft == 0)
+}
+
+// NonDefaultRoute returns whether the passed Route is the default
+func NonDefaultRoute(route netlink.Route) bool {
+	return route.Dst != nil
+}
+
+func ContainedByCIDR(cidr string) AddressFilter {
+	return func(addr netlink.Addr) bool {
+		_, parsedNet, err := net.ParseCIDR(cidr)
+		if err != nil {
+			panic(err)
+		}
+		return parsedNet.Contains(addr.IP)
+	}
+}
+
+func AddressNotIn(ips ...string) AddressFilter {
+	return func(addr netlink.Addr) bool {
+		for _, ip := range ips {
+			if addr.IP.String() == ip {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+func AddressFilters(filters ...AddressFilter) AddressFilter {
+	return func(addr netlink.Addr) bool {
+		for _, include := range filters {
+			if !include(addr) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+type addrMap map[netlink.Link][]netlink.Addr
+type routeMap map[int][]netlink.Route
+
+// routableAddresses takes a slice of Virtual IPs and returns a slice of
+// configured addresses in the current network namespace that directly route to
+// those vips. You can optionally pass an AddressFilter and/or RouteFilter to
+// further filter down which addresses are considered.
+//
+// This is ported from https://github.com/openshift/baremetal-runtimecfg/blob/master/pkg/utils/utils.go.
+func routableAddresses(addrMap addrMap, routeMap routeMap, vips []net.IP, af AddressFilter, rf RouteFilter) ([]net.IP, error) {
```

Comment:
>I think it is in all of our best interests if this were part of some shared library. at a min we should TODO that. I forsee a bug they patch and we don't etc.

It would be useful to enumerate every use case for it outside this repo to justify doing so.

>Do we have any perf concerns with all of these nested loops I mean I guess its a result of the complexity of the problem. But I am curious what render looks like on a box with a bunch of interfaces. Does the kernel or something else do this work that we can ask?

On a box with 6 interfaces the command completes apparently "instantaneously" although I don't have the wall clock measurement from the box I was using which is long gone. I can't really conceive of any performance concerns in this context.

Reply:
thanks for looking